### PR TITLE
Less allocs in resid2DLinear and faster mean

### PR DESCRIPTION
**Pull request by Affie**

### CODE REVIEW

File: src/CommonUtils.jl [hunk 1]
```diff
@@ -2,10 +2,11 @@
 
 
 function resid2DLinear(μ, mus, Lambdas; diffop::Function=-)  # '-' exploits EuclideanManifold commutativity a-b = b-a
-  dμ = broadcast(diffop, μ, mus)  # mus .- μ  ## μ .\ mus
+  # dμ = broadcast(diffop, μ, mus)  # mus .- μ  ## μ .\ mus
   # @show round.(dμ, digits=4)
-  ret = sum( Lambdas.*dμ )
-  return ret
+  # ret = sum( Lambdas.*dμ )
+  r = map((mu, lam) -> diffop(μ[], mu) * lam, mus, Lambdas)
```
Comment (Affie) [inline src/CommonUtils.jl:8]: @dehann, here is a version without the extra dispatch and allocations. 
Is μ always of length 1?

File: src/CommonUtils.jl [hunk 2]
```diff
@@ -71,7 +72,7 @@ function _getManifoldFullOrPart(mkd::ManifoldKernelDensity, aspartial::Bool=true
 end
 
 function Statistics.mean(mkd::ManifoldKernelDensity, aspartial::Bool=true; kwargs...)
-  return mean(_getManifoldFullOrPart(mkd,aspartial), getPoints(mkd, aspartial); kwargs...)
+  return mean(_getManifoldFullOrPart(mkd,aspartial), getPoints(mkd, aspartial), GeodesicInterpolation(); kwargs...)
```
Comment (Affie) [inline src/CommonUtils.jl:75]: GeodesicInterpolation is currently way faster but still has dynamic dispatch.

Comment (mateuszbaran) [reply src/CommonUtils.jl:75]: `ExtrinsicEstimation` should usually be the fastest way to compute `mean`. It doesn't work well for some manifolds but for Lie groups it is worth trying out when you prefer speed over accuracy.

**Comment by Affie** [comment]
```julia
fg = generateGraph_Hexagonal()
@time solveGraph!(fg; multithread=true)
# 2.072770 seconds (16.03 M allocations: 742.617 MiB, 4.83% gc time)

@time solveGraph!(fg)
# 4.217847 seconds (15.65 M allocations: 717.741 MiB, 5.96% gc time)
```

File: src/CommonUtils.jl (hunk 2 above)
Comment (Affie) [reply src/CommonUtils.jl:75]: Thanks, I'll give it a try. With all the performance enhancements you helped with so far we are seeing a major improvement already.

Comment (mateuszbaran) [reply src/CommonUtils.jl:75]: Great! Feel free to report any further issues you think I could help with.

Comment (Affie) [reply src/CommonUtils.jl:75]: `ExtrinsicEstimation` gives an error:
```
MethodError: no method matching get_embedding(::GroupManifold{ℝ, ProductManifold{ℝ, Tuple{TranslationGroup{ManifoldsBase.TypeParameter{Tuple{2}}, ℝ}, SpecialOrthogonal{ManifoldsBase.TypeParameter{Tuple{2}}}}}, Manifolds.SemidirectProductOperation{RotationAction{LeftAction, TranslationGroup{ManifoldsBase.TypeParameter{Tuple{2}}, ℝ}, SpecialOrthogonal{ManifoldsBase.TypeParameter{Tuple{2}}}}}})
```
Don't worry about it now though. I think there are other optimizations that will make a bigger difference.

File: src/CommonUtils.jl (hunk 1 above)
Comment (dehann) [reply src/CommonUtils.jl:8]: I think so, will double check bit later today.

Comment (dehann) [reply src/CommonUtils.jl:8]: looks right

**Review by dehann — APPROVED** [review]

**Pull request merged**